File: EvoBio4/Collections/CooperatorGroup.cs
using System.Collections.Generic;
using EvoBio4.Core.Enums;
using EvoBio4.Core.Interfaces;

namespace EvoBio4.Collections
{
	public class CooperatorGroup : IndividualGroupBase, ICooperatorGroup<Individual>
	{
		public List<Individual> ReproducingIndividuals { get; private set; }
		public List<Individual> NonReproducingIndividuals { get; private set; }
		public double ReproducingQualitySum { get; private set; }
		public double NonReproducingQualitySum { get; private set; }

		public CooperatorGroup ( ) :
			base ( IndividualType.Cooperator )
		{
		}

		public void Split ( double threshold )
		{
			ReproducingIndividuals    = new List<Individual> ( Individuals.Count );
			NonReproducingIndividuals = new List<Individual> ( Individuals.Count );
			ReproducingQualitySum     = 0;
			NonReproducingQualitySum  = 0;

			foreach ( var individual in Individuals )
				if ( individual.Quality < threshold )
				{
					NonReproducingIndividuals.Add ( individual );
					NonReproducingQualitySum += individual.Quality;
				}
				else
				{
					ReproducingIndividuals.Add ( individual );
					ReproducingQualitySum += individual.Quality;
				}

			QualitySum = ReproducingQualitySum + NonReproducingQualitySum;
		}
	}
}

File: EvoBio4.Cpp/cooperator_group.cpp
#include "cooperator_group.h"


cooperator_group::cooperator_group () : individual_group (IT_COOPERATOR) {}

cooperator_group::~cooperator_group () = default;

void cooperator_group::split (double threshold)
{
	reproducing.clear ();
	non_reproducing.clear ();
	reproducing.reserve (individuals.size ());
	non_reproducing.reserve (individuals.size ());
	quality_of.reproducing = 0;
	quality_of.non_reproducing = 0;

	for (auto&& individual : individuals)
	{
		if (individual->quality < threshold)
		{
			non_reproducing.emplace_back (individual);
			quality_of.non_reproducing += individual->quality;
		}
		else
		{
			reproducing.emplace_back (individual);
			quality_of.reproducing += individual->quality;
		}
	}

	quality = quality_of.reproducing + quality_of.non_reproducing;
}
